examples/dummy_plugin/dummy_plugin/api/views.py

```diff
@@ -9,3 +13,21 @@ class DummyViewSet(ModelViewSet):
     queryset = DummyModel.objects.all()
     serializer_class = DummyModelSerializer
     filterset_class = DummyModelFilterSet
+
+
+#
+# Webhook Testing
```

Comment:
It feels wrong to me to have this in the dummy plugin; as IMO that should stay as a clean example of how to write a plugin. This is test code - did the approach we discussed the other day about defining the test view and urlpatterns inside the `tests` directory and using `@override_settings(ROOT_URLCONF...)` not work out for some reason?

Reply:
Becuase of the generic classes we use it looks like we relie on other urls being defined. When we set `@override_settings(ROOT_URLCONF...)` we remove all the other URL paths. At the moment I think it would be very difficult to define all the required paths because of the generic classes we use.

`django.urls.exceptions.NoReverseMatch: 'plugins-api' is not a registered namespace`

`django.core.exceptions.ImproperlyConfigured: Could not resolve URL for hyperlinked relationship using view name "plugins-api:dummy_plugin-api:dummymodel-detail". You may have failed to include the related model in your API, or incorrectly configured the `lookup_field` attribute on this field.`

Reply:
Not sure I follow - with this as a self-contained `SplinterTestCase`, I would have thought the only URL it should require is the webhook endpoint that it's explicitly using? If it were a `ViewTestCases` or an `APIViewTestCases` subclass, sure, I could see that being an issue. But the only HTTP call we should be making in this test case is the webhook one.

Reply:
Any further comments here?

Reply:
No doesnt work like that, when I change the `ROOL_URL` it raises an exception for `plugins-api` not being registered as it wipes out all other URLs from Nautobot. In one of the generic views there is a lookup to `plugins-api` causing the test to fail.

Reply:
```
======================================================================
ERROR: test_plugin_webhook_update (nautobot.extras.tests.integration.test_plugins.PluginWebhookTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/django/urls/base.py", line 72, in reverse
    extra, resolver = resolver.namespace_dict[ns]
KeyError: 'plugins-api'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/rest_framework/relations.py", line 401, in to_representation
    url = self.get_url(value, self.view_name, request, format)
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/rest_framework/relations.py", line 339, in get_url
    return self.reverse(view_name, kwargs=kwargs, request=request, format=format)
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/rest_framework/reverse.py", line 47, in reverse
    url = _reverse(viewname, args, kwargs, request, format, **extra)
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/rest_framework/reverse.py", line 60, in _reverse
    url = django_reverse(viewname, args=args, kwargs=kwargs, **extra)
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/django/urls/base.py", line 83, in reverse
    raise NoReverseMatch("%s is not a registered namespace" % key)
django.urls.exceptions.NoReverseMatch: 'plugins-api' is not a registered namespace

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kristian/projects/nautobot/nautobot/nautobot/extras/tests/integration/test_plugins.py", line 115, in test_plugin_webhook_update
    obj.validated_save()
  File "/Users/kristian/projects/nautobot/nautobot/nautobot/core/models/__init__.py", line 52, in validated_save
    self.save()
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/django/db/models/base.py", line 754, in save
    force_update=force_update, update_fields=update_fields)
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/django/db/models/base.py", line 803, in save_base
    update_fields=update_fields, raw=raw, using=using,
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/django/dispatch/dispatcher.py", line 179, in send
    for receiver in self._live_receivers(sender)
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/django/dispatch/dispatcher.py", line 179, in <listcomp>
    for receiver in self._live_receivers(sender)
  File "/Users/kristian/projects/nautobot/nautobot/nautobot/utilities/utils.py", line 311, in _curried
    return _curried_func(*args, *moreargs, **{**kwargs, **morekwargs})
  File "/Users/kristian/projects/nautobot/nautobot/nautobot/extras/signals.py", line 67, in _handle_changed_object
    enqueue_webhooks(instance, request.user, request.id, action)
  File "/Users/kristian/projects/nautobot/nautobot/nautobot/extras/webhooks.py", line 43, in enqueue_webhooks
    serializer.data,
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/rest_framework/serializers.py", line 548, in data
    ret = super().data
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/rest_framework/serializers.py", line 246, in data
    self._data = self.to_representation(self.instance)
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/rest_framework/serializers.py", line 515, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/Users/kristian/Library/Caches/pypoetry/virtualenvs/nautobot-QbOQWJa6-py3.7/lib/python3.7/site-packages/rest_framework/relations.py", line 416, in to_representation
    raise ImproperlyConfigured(msg % self.view_name)
django.core.exceptions.ImproperlyConfigured: Could not resolve URL for hyperlinked relationship using view name "plugins-api:dummy_plugin-api:dummymodel-detail". You may have failed to include the related model in your API, or incorrectly configured the `lookup_field` attribute on this field.
```

Reply:
OK, I understand now. `enqueue_webhooks` uses the REST API serializer to encode the object, and the serializer includes the REST API URL of the object in question, so if we don't have the REST API URL patterns available, it errors out as above. I'm okay with keeping this in the example plugin for the time being, though perhaps we can open a housekeeping item to investigate refactoring this to somewhere else (perhaps a dedicated "tests" plugin or such, or figuring out how to augment rather than overwrite the available urlpatterns for testing?)